clang/test/CodeGen/fp-accuracy.c

```diff
@@ -1,46 +1,67 @@
+// RUN: "-ffp-builtin-accuracy=high:[sin,cosf] low:[tan] medium:[sincos]" \
 // RUN: -Wno-implicit-function-declaration -emit-llvm -o - %s \
-// RUN: | FileCheck --check-prefix=CHECK-FUNC %s
+// RUN: | FileCheck --check-prefix=CHECK-FUNC-1 %s
 
-// RUN: %clang_cc1 -triple spir64-unknown-unknown \
-// RUN: -D SPIR -ffp-accuracy-attr=sycl \
+// RUN: %clang_cc1 -triple x86_64-unknown-unknown \
+// RUN: "-ffp-builtin-accuracy=medium high:[tan] cuda:[cos]" \
+// RUN: -Wno-implicit-function-declaration -emit-llvm -o - %s \
+// RUN: | FileCheck --check-prefix=CHECK-FUNC-2 %s
+
+// RUN: %clang_cc1 -triple spir64-unknown-unknown -ffp-builtin-accuracy=sycl \
 // RUN: -Wno-implicit-function-declaration -emit-llvm -o - %s \
 // RUN: | FileCheck --check-prefix=CHECK-SPIR %s
 
-float a, b, c, d, e, f;
+float a, b, c, d, f;
+double e;
+
+extern double sincos (double __x) __attribute__ ((__nothrow__ ));
 
 void foo(void) {
   // CHECK-LABEL: define {{.*}}void @foo()
   a = cosf(b);
   c = sin(a);
   d = tan(c);
   e = cos(d);
-#ifndef SPIR
   d = sincos(e);
-#endif
-  // CHECK: call float @llvm.fpbuiltin.cos.f32(float {{.*}})        [[ATTR3:#[0-9]+]]
-  // CHECK: call double @llvm.fpbuiltin.sin.f64(double {{.*}})      [[ATTR3:#[0-9]+]]
-  // CHECK: call double @llvm.fpbuiltin.tan.f64(double {{.*}})      [[ATTR3:#[0-9]+]]
-  // CHECK: call double @llvm.fpbuiltin.cos.f64(double {{.*}})      [[ATTR3:#[0-9]+]]
-
-  // CHECK-FUNC: call float @llvm.fpbuiltin.cos.f32(float {{.*}})   [[ATTR3:#[0-9]+]]
-  // CHECK-FUNC: call double @llvm.fpbuiltin.sin.f64(double {{.*}}) [[ATTR3:#[0-9]+]]
-  // CHECK-FUNC: call double @llvm.fpbuiltin.tan.f64(double {{.*}}) [[ATTR4:#[0-9]+]]
-  // CHECK-FUNC: call double @llvm.fpbuiltin.cos.f64(double {{.*}})
-  // CHECK-FUNC: call i32 (double, ...) @sincos(double {{.*}})      [[ATTR5:#[0-9]+]]
-
-  // CHECK-SPIR: call float @llvm.fpbuiltin.cos.f32(float {{.*}}    [[ATTR5:#[0-9]+]]
-  // CHECK-SPIR: call double @llvm.fpbuiltin.sin.f64(double {{.*}}) [[ATTR5:#[0-9]+]]
-  // CHECK-SPIR: call double @llvm.fpbuiltin.tan.f64(double {{.*}}) [[ATTR6:#[0-9]+]]
-  // CHECK-SPIR: call double @llvm.fpbuiltin.cos.f64(double {{.*}}  [[ATTR5:#[0-9]+]]
+
+  // CHECK: call float @llvm.fpbuiltin.cos.f32(float {{.*}})        [[ATTR_HIGH:#[0-9]+]]
+  // CHECK: call double @llvm.fpbuiltin.sin.f64(double {{.*}})      [[ATTR_HIGH]]
+  // CHECK: call double @llvm.fpbuiltin.tan.f64(double {{.*}})      [[ATTR_HIGH]]
+  // CHECK: call double @llvm.fpbuiltin.cos.f64(double {{.*}})      [[ATTR_HIGH]]
+
+  // CHECK-FUNC-1: call float @llvm.fpbuiltin.cos.f32(float {{.*}})   [[ATTR_HIGH:#[0-9]+]]
+  // CHECK-FUNC-1: call double @llvm.fpbuiltin.sin.f64(double {{.*}}) [[ATTR_HIGH]]
+  // CHECK-FUNC-1: call double @llvm.fpbuiltin.tan.f64(double {{.*}}) [[ATTR_LOW:#[0-9]+]]
+  // CHECK-FUNC-1: call double @llvm.cos.f64(double {{.*}})
+  // CHECK-FUNC-1: call double @sincos(double {{.*}})                 [[ATTR_MEDIUM:#[0-9]+]]
```

Comment:
I hadn't noticed this before. This needs to be a call to @llvm.fpbuiltin.sincos.f64(). I guess that's not happening because sincos isn't a clang builtin so it doesn't get the same handling as sin, cos, and tan. We need to make this replacement somewhere.

Reply:
The intrinsic is defined here: https://github.com/intel/llvm/blob/sycl/llvm/include/llvm/IR/Intrinsics.td#L891
but it requires 2 arguments. And as far as I know it takes only one?

Reply:
No, it's a weird function. My current definition is wrong though. It actually takes three arguments -- the input value, an output pointer for sin, and an output pointer for cos.

https://man7.org/linux/man-pages/man3/sincos.3.html